### ISSUE TITLE
Should assignment to parameters greedily move value?

### ISSUE THREAD

**Issue by springmeyer** [mapnik/mapnik#2651]
@artemp: is this expected behavior?

```cpp
std::string tmp("rvalue");
mapnik::parameters params;
params["rvalue"] = tmp;
// now tmp is empty since it has been moved
````

My sense is that this is okay, but slightly dangerous in the case where I noticed it. I did:

```cpp
params["key_field"] = key_field_;
```

In the postgis plugin constructor and then `key_field_` was modified creating havoc later on. I fixed this by doing:

```cpp
params["key_field"] = std::string(key_field_);
```

To trigger a copy and avoid `key_field_` being modified. Thoughts?

**Comment by artemp** [comment]
@springmeyer - not intentional. In fact both copy and move semantics should work as follows 

* use copy ctor by default
* use move ctor when value is R-value (either explicit or marked with std::move(..)) 

So something is not working here.

**Comment by springmeyer** [comment]
Okay, thanks for having a look: maybe universal references are involved?

**Comment by artemp** [comment]
fixed in  1f3e36c3ec260ab9fd8c798e22362b65ca694eb9

```c++
    std::string str("value");
    std::cerr << "#1 " << str << std::endl; // original value
    mapnik::parameters params;
    params["value"] = str; // copy
    std::cerr << "#2 " << str << std::endl; // str is untouched
    params["value"] = std::move(str);
    std::cerr << "#3 " << str << std::endl; // str moved 
```
